ResearchKit/Common/ORKPermissionsValidator.m

```diff
@@ -0,0 +1,112 @@
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#import "ORKPermissionsValidator.h"
+
+NSString * const ORK_NSMotionUsageDescription = @"NSMotionUsageDescription";
+NSString * const ORK_NSMicrophoneUsageDescription = @"NSMicrophoneUsageDescription";
+NSString * const ORK_NSLocationWhenInUseUsageDescription = @"NSLocationWhenInUseUsageDescription";
+NSString * const ORK_NSLocationAlwaysUsageDescription = @"NSLocationAlwaysUsageDescription";
+NSString * const ORK_NSCameraUsageDescription = @"NSCameraUsageDescription";
+NSString * const ORK_NSHealthShareUsageDescription = @"NSHealthShareUsageDescription";
+NSString * const ORK_NSHealthUpdateUsageDescription = @"NSHealthUpdateUsageDescription";
+
+@implementation ORKPermissionsValidator
+
++ (NSString *)findBundleKey:(NSString*)key error:(NSError **)error {
+    
+    // Look for the string in the info.plist and if not found then return error
+    NSDictionary *info = [[NSBundle mainBundle] localizedInfoDictionary] ? : [[NSBundle mainBundle] infoDictionary];
+    NSString *str = info[key];
+    
+    // Create the error if not found
+    if (![str isKindOfClass:[NSString class]] && error) {
+        NSString *reason = [NSString stringWithFormat:@"Missing required Info.plist key: %@", key];
+        NSString *description = [NSString stringWithFormat:@"'%@' is a required key for the authorizations you have requested. See https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/CocoaKeys.html", key];
+        *error = [NSError errorWithDomain:NSCocoaErrorDomain code:1 userInfo:@{NSLocalizedFailureReasonErrorKey : reason,
+                                                                               NSLocalizedDescriptionKey : description}];
+        str = nil;
+    }
+    
+    return str;
+}
+
++ (BOOL)validatePermissions:(ORKPermissionMask)permissions
+                 writeTypes:(nullable NSSet<HKSampleType *> *)typesToWrite
+                  readTypes:(nullable NSSet<HKObjectType *> *)typesToRead
+                      error:(NSError **)error {
+    
+    // NOTE: if statements are intentionally nested for clarity where if the permission isn't
+    // included, then don't look for the key in the Info.plist (or any other validation that may
+    // be useful to include here)
+    
+    if ((permissions & ORKPermissionCoreMotionActivity) ||
+        (permissions & ORKPermissionCoreMotionAccelerometer)) {
+        if (![self findBundleKey:ORK_NSMotionUsageDescription error:error]) {
+            return NO;
+        }
+    }
+    
+    if ((permissions & ORKPermissionAudioRecording)) {
+        if (![self findBundleKey:ORK_NSMicrophoneUsageDescription error:error]) {
+            return NO;
+        }
+    }
+    
+    if ((permissions & ORKPermissionCoreLocation)) {
+        if (![self findBundleKey:ORK_NSLocationWhenInUseUsageDescription error:error] &&
```

Comment:
Is it not possible for a RK app to only request location when in use? Or does iOS still require the "always" description even if you only ever ask for "when in use" permission?